packages/state-transition/src/util/loadState.ts

```diff
@@ -0,0 +1,251 @@
+    isMoreValidator ? oldInactivityScores : oldInactivityScores.subarray(0, minValidator * 8),
+    isMoreValidator ? inactivityScoresBytes.subarray(0, minValidator * 8) : inactivityScoresBytes,
+    modifiedValidators
+  );
+
+  for (const validatorIndex of modifiedValidators) {
+    migratedState.inactivityScores.set(
+      validatorIndex,
+      ssz.UintNum64.deserialize(inactivityScoresBytes.subarray(validatorIndex * 8, (validatorIndex + 1) * 8))
+    );
+  }
+
+  if (isMoreValidator) {
+    // add new inactivityScores
+    for (let validatorIndex = oldValidator; validatorIndex < newValidator; validatorIndex++) {
+      migratedState.inactivityScores.push(
+        ssz.UintNum64.deserialize(inactivityScoresBytes.subarray(validatorIndex * 8, (validatorIndex + 1) * 8))
+      );
+    }
+  } else {
+    if (newValidator - 1 < 0) {
+      migratedState.inactivityScores = ssz.altair.InactivityScores.defaultViewDU();
+    } else {
+      migratedState.inactivityScores = migratedState.inactivityScores.sliceTo(newValidator - 1);
+    }
+  }
+}
+
+function loadValidators(
+  migratedState: BeaconStateAllForks,
+  seedState: BeaconStateAllForks,
+  newValidatorsBytes: Uint8Array
+): number[] {
+  const seedValidatorCount = seedState.validators.length;
+  const newValidatorCount = Math.floor(newValidatorsBytes.length / VALIDATOR_BYTES_SIZE);
+  const isMoreValidator = newValidatorCount >= seedValidatorCount;
+  const minValidatorCount = Math.min(seedValidatorCount, newValidatorCount);
+  // migrated state starts with the same validators to seed state
+  migratedState.validators = seedState.validators.clone();
+  const seedValidatorsBytes = seedState.validators.serialize();
+  const modifiedValidators: number[] = [];
+  findModifiedValidators(
+    isMoreValidator ? seedValidatorsBytes : seedValidatorsBytes.subarray(0, minValidatorCount * VALIDATOR_BYTES_SIZE),
+    isMoreValidator ? newValidatorsBytes.subarray(0, minValidatorCount * VALIDATOR_BYTES_SIZE) : newValidatorsBytes,
+    modifiedValidators
+  );
+  for (const i of modifiedValidators) {
+    migratedState.validators.set(
+      i,
+      ssz.phase0.Validator.deserializeToViewDU(
+        newValidatorsBytes.subarray(i * VALIDATOR_BYTES_SIZE, (i + 1) * VALIDATOR_BYTES_SIZE)
+      )
+    );
+  }
+
+  if (newValidatorCount >= seedValidatorCount) {
+    // add new validators
+    for (let validatorIndex = seedValidatorCount; validatorIndex < newValidatorCount; validatorIndex++) {
+      migratedState.validators.push(
+        ssz.phase0.Validator.deserializeToViewDU(
+          newValidatorsBytes.subarray(
+            validatorIndex * VALIDATOR_BYTES_SIZE,
+            (validatorIndex + 1) * VALIDATOR_BYTES_SIZE
+          )
+        )
+      );
+      modifiedValidators.push(validatorIndex);
+    }
+  } else {
+    migratedState.validators = migratedState.validators.sliceTo(newValidatorCount - 1);
+  }
+  return modifiedValidators;
+}
+
+function findModifiedValidators(
```

Comment:
This function performs a binary search splitting the serialized validators buffer finding differences. Can you prove this is faster than just comparing byte slices? i.e. a naive approach of comparing each validator one by one. As optimization target you can write custom comparison code to check equality of a serialized validator against a de-serialized struct. That would prevent having to allocate more memory for the full buffer to buffer comparison.

Reply:
If this function remains, please write a nice comment above with JSDoc syntax `/**`